sfge/src/main/java/com/salesforce/rules/unusedmethod/RuleStateTracker.java

```diff
@@ -196,11 +210,174 @@ List<String> getSubclasses(String... definingTypes) {
+                                        String outerType = v.getDefiningType().split("\\.")[0];
+                                        // Get any inner classes for the outer type and cast their
+                                        // names to lowercase.
+                                        Set<String> innerClassNames =
+                                                getInnerClasses(outerType).stream()
+                                                        .map(i -> i.getName().toLowerCase())
+                                                        .collect(Collectors.toSet());
+                                        // If the set lacks an entry for our referenced type, then
+                                        // there's no conflicting inner class and we can keep this
+                                        // result.
+                                        return !innerClassNames.contains(
+                                                referencedType.toLowerCase());
+                                    })
+                            .collect(Collectors.toList());
+        }
+        return results;
+    }
+
+    /**
+     * Returns the {@link BaseSFVertex} where the value referenced in the specified method call is
+     * declared. <br>
+     * E.g., for {@code someObject.someMethod()}, returns the declaration of {@code someObject}.
+     */
+    Optional<BaseSFVertex> getDeclarationOfReferencedValue(MethodCallExpressionVertex methodCall) {
+        // Step 1: Get the name of the thing being referenced.
+        List<String> referenceNameList = methodCall.getReferenceExpression().getNames();
+        // Step 2: The method call must happen in the context of another method, a field
+        // declaration, or both. Determine which.
+        Optional<MethodVertex> parentMethodOptional = methodCall.getParentMethod();
+        Optional<FieldDeclarationVertex> parentFieldDeclarationOptional =
+                methodCall.getFieldDeclaration();
+        // Theoretically one or both of those optionals should be present. If not, throw an
+        // exception.
+        if (!parentMethodOptional.isPresent() && !parentFieldDeclarationOptional.isPresent()) {
+            throw new UnexpectedException(
+                    "Cannot determine context for method call " + methodCall.toMinimalString());
+        }
+        // Step 3: If we're in a method, check for variables and parameters.
+        if (parentMethodOptional.isPresent()) {
+            MethodVertex parentMethod = parentMethodOptional.get();
+            // Step 3A: Check for variables.
+            List<VariableDeclarationVertex> declaredVariables =
+                    MethodUtil.getVariableDeclarations(g, parentMethod);
+            for (VariableDeclarationVertex declaredVariable : declaredVariables) {
+                // NOTE: A known bug exists here. We merely check that the variable exists,
```

Comment:
Calling this out explicitly so we can have a discussion. This implementation is a little naive, and the following cases will lead to false positives:
```
SomeClass someClass = SomeClass.getInstance();
```
```
Boolean b = SomeClass.getBoolean();
SomeClass someClass = new SomeClass();
```
```
if (x) {
Boolean b = SomeClass.getBoolean();
} else {
SomeClass someClass = new SomeClass();
}
```
However, since Apex is case-insensitive, naming a variable after a class can often cause compile errors anyway. Therefore, it's my opinion that this is an acceptable limitation, and we're better off implementing a Best Practices rule that prevents variables from being named after classes, instead of expending the effort to support cases like this with UnusedMethodRule.
Thoughts?

Reply:
Sounds reasonable though I'm not sure we need Graph Engine to create a rule like this and what the accepted Apex naming conventions are.
If the vertex implements `Typeable`, we should be able to look up the type information. Can we put a log line to ignore these cases? Or a variation of the violation message to indicate that this could be a false positive?

Reply:
`VariableDeclarationVertex` implements `Typeable`.

Reply:
The specific problem is less about the `VariableDeclarationVertex`'s type information and more simply about the fact that the variable shares the same name as the class.
In a case where the variable is declared before the method call, then the method call is on that variable instead of the static class. The false positive case is when the variable is declared later, because in that case the method call will still be assumed (incorrectly) to be on that variable.
We could theoretically make a "this might be a false positive" variant of the error message, but I'm honestly inclined to just leave it as is. We can reassess if we get any complaints, but it feels like a lot of effort to solve a problem we don't know for a fact anyone will actually have.

Reply:
Agree. Let's wait and see.